### PR TITLE
Migration to .NET Standard

### PR DESCRIPTION
**Pull request by vladapokorny**
- Update NuGet packages
- Upgrade CSPROJ files to new format
- Upgrade projects to .NET Standard
- Build without Rake

### CODE REVIEW

**Comment by haf** [comment]
Ready to merge?

**Comment by vladapokorny** [comment]
One main change is that Rake build is currently completely disabled, because current configuration is incompatible with new csproj, so the solution is changed to be capable build without any third party tool including building NuGet packages. The main disadvantage is that version number must be manually updated before build and unit tests also must be started manully.

On the other hand, projects currently targets to .NET Standard 2.0 and unit tests are passing, so I think solution is ready to merge.

**Comment by haf** [comment]
Ok, I think @jonorossi or someone has removed my write access, but for me: LGTM.

**Comment by jonorossi** [comment]
@vladapokorny thanks, I'll take a look at it now. This project was on my list to move to our deprecated organisation, the last release was in 2012 and it is unmaintained, so glad to hear someone is still making use of it and that it might stick around.

@haf see https://groups.google.com/d/topic/castle-project-devel/7YoT127yZdg/discussion for the details just over a year ago.

**Comment by haf** [comment]
@jonorossi Yeah, I'm fine not being active, and I don't keep track of that list any more; too many sources of inputs for me these days...

**Comment by jonorossi** [comment]
@haf no problem.

Could you please add the `castleproject` user to the Castle Transactions nuget.org packages, it appears many of the packages mentioned in the readme don't even exist on nuget.org.

File: Castle.Transactions.sln
Comment (jonorossi) [inline Castle.Transactions.sln]: You can remove all the other build configurations, since you upgraded to .NET 4.6.1.

**Comment by haf** [comment]
Sure. Done.

File: src/Castle.Facilities.AutoTx.Tests/Castle.Facilities.AutoTx.Tests.csproj
Comment (jonorossi) [inline src/Castle.Facilities.AutoTx.Tests/Castle.Facilities.AutoTx.Tests.csproj]: Based on your description I expected to see only .NET SDK based projects. I'd have a look at Castle Core or Windsor how our maintained projects are being done.

File: src/Castle.Facilities.AutoTx.Tests/Properties/AssemblyInfo.cs
Comment (jonorossi) [inline src/Castle.Facilities.AutoTx.Tests/Properties/AssemblyInfo.cs]: As mentioned elsewhere it would be great to get this project working like Castle Core and Windsor which automatically pick up their version from Git tags.

File: src/Castle.Transactions/Activities/ThreadLocalActivityManager.cs
Comment (jonorossi) [inline src/Castle.Transactions/Activities/ThreadLocalActivityManager.cs]: You've renamed `CallContextActivityManager` to `ThreadLocalActivityManager` and obviously changed the implementation, this appears to be a backwards change as the recommendation is for things to use the call context and move away from thread local. Could you explain this, and update the XML comment.

Also your indentation is off, you probably want to copy editorconfig from Castle Core.

File: Gemfile
Comment (jonorossi) [inline Gemfile]: I think it is time to remove ruby from the build process. With the .NET SDK tooling performing most of the work we've been able to do the few additional things as part of that process with Castle Core and Windsor.

**Review by jonorossi — CHANGES_REQUESTED** [review]
Quite a few review comments since this project hasn't really been touched since 2012 and is in a pretty poor state since tooling has changed quite a lot since then.

We also need to get a build set up, we are using AppVeyor for other Castle projects. We will need to look at packaging because this project is considerably more complicated with Castle.IO and others libraries which weren't part of the original Castle Transactions and appear not to have been released.

There is quite a bit of work to overhaul this to get it released, I wanted to outline those things to make sure you are committed.

File: Gemfile
Comment (haf) [reply Gemfile]: I think he wanted to separate PRs; this PR does remove the build's dependency on it. Another can remove it.

**Comment by vladapokorny** [comment]
I made some updates (cleanup SLN file, remove Rake build process, finish migration of unit tests to new csproj format).

CallContextActivityManager was replaced with ThreadLocalActivityManager, because .NET Standard 2.0 doesn't contain CallContext class. My first idea was replace CallContext with AsyncLocal, but this step has broken unit tests. Maybe only problem is that unit test are not ready for AsyncLocal, but I didn't investigate this.

Build process migration is another chapter. I didn't know which way is correct, so I migrated it only to standard MSBuild. I haven't worked with AppVeyor yet and I don't have currently much time to examine how it works.

**Comment by fir3pho3nixx** [comment]
@vladapokorny can I urge you to reconsider your position on AsyncLocal. We are using it in Castle Windsor with great success. It is the right way to go.

**Comment by vladapokorny** [comment]
@fir3pho3nixx I agree that AsyncLocal will be better. As I mentioned, the problem is with unit tests. If ThreadLocal is replaced with AsyncLocal then the following tests fail (or never finish):

- MultipleThreads_TransactionBookKeeping/Count_InTheAmbientContext_IsOne
- MultipleThreads_TransactionBookKeeping/InterleavingWhereChildCommit_IsOutstanding_OverParent
- PerTransactionLifestyle_Releasing/Concurrent_DependentTransaction_AndDisposing

**Comment by hconceicao** [comment]
The issue is a little more complicated than some unit tests failing.

In real world scenarios involving threadpools and schedulers (asp.net requests or tpl for example), there is no guarantee that a transaction will be executed on the same thread. The runtime can pause the invocation and resume on another thread, and a thread local store will mix data and etc.


Cheers,
Henry Conceição

**Comment by fir3pho3nixx** [comment]
@vladapokorny, we should have a closer look at those tests. Are they still valid? Perhaps they are asserting the right thing but in the wrong way. It is a fight worth having. Do you have something on a branch you tried so I can have a look?

**Comment by vladapokorny** [comment]
@fir3pho3nixx I created AsyncLocal branch with ThreadAsync replaced with AsyncLocal and enabled logging (https://github.com/Misterinecompany/Castle.Transactions/tree/AsyncLocal). Three tests from Castle.Facilities.AutoTx.Tests are failing (one fails directly, the two others never end – exception is thrown, so ManualResetEvent never receives signal to continue).
I’m not much familiar with inner logic, so I haven't done any other steps.

**Comment by fir3pho3nixx** [comment]
@vladapokorny 

Please see: https://github.com/Misterinecompany/Castle.Transactions/pull/1

The tests don't seem to run in Visual Studio. This is because we are missing "magic" package references! :)

The first test that jumped out at me was:

image

If we need single apartment threading, then perhaps would investigate upgrading the tests to make sure none of them run in parallel. 

Something like: https://github.com/nunit/docs/wiki/SingleThreaded-Attribute

This will get us started. What do you think?

**Comment by vladapokorny** [comment]
Good job. I have run tests using ReSharper, so I didn't faced with this problem. Running Castle.Transactions.Tests in STA is now resolved by using https://github.com/nunit/docs/wiki/Apartment-Attribute

**Comment by fir3pho3nixx** [comment]
@vladapokorny 

Please see my [PR](https://github.com/Misterinecompany/Castle.Transactions/pull/2). This fixed the tests in Castle.Transactions.Tests. 👍 

After digging into the tests a bit more, I noticed the following about [MultipleThreads_TransactionBookKeeping.Count_InTheAmbientContext_IsOne](https://github.com/castleproject/Castle.Transactions/blob/master/src/Castle.Facilities.AutoTx.Tests/MultipleThreads_TransactionBookKeeping.cs#L38). 

image

It would seem the transitive cloned transaction from a forked transaction based on ambient state is causing the failure here. 

Does anyone involved in this PR think it might be a problem with System.Transactions.Transaction.DependentClone on netstandard?

**Comment by fir3pho3nixx** [comment]
We need to keep going on this. The formula we are using for this in Windsor is:

 - net45 = [System.Runtime.Remoting.Messagin.CallContext](https://github.com/castleproject/Windsor/blob/master/src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/CallContextLifetimeScope.cs#L123)
 - netstandard1.6 = [System.Threading.AsyncLocal](https://github.com/castleproject/Windsor/blob/master/src/Castle.Windsor/MicroKernel/Lifestyle/Scoped/CallContextLifetimeScope.cs#L126)

What do you think?

**Comment by vladapokorny** [comment]
@fir3pho3nixx
1. It is good, but remaining tests are more problematic.
2. Two profiles for net45 and netstandard2.0 (maybe netstandard1.6) are possible, but I think that we need to resolve unit test with AsyncLocal primarily. Another question is if it is still required to support .NET Framework 4.5.

I'm going to look at these tests a little bit more.

**Comment by vladapokorny** [comment]
@fir3pho3nixx OK, let's focus on AsyncLocal. I analyzed tests deeper. MultipleThreads_TransactionBookKeeping.Count_InTheAmbientContext_IsOne is harder to completely analyze because of multiple recursion calls, so I focused more on PerTransactionLifestyle_Releasing.Concurrent_DependentTransaction_AndDisposing and my observation is that AsyncLocal has different behavior than CallContext.GetData (probably CallContext.LogicalGetData is closer to AsyncLocal).

The result is that unit tests assume that starting new Task creates a new context, so CallContext doesn't contain any Activity instance for a new context, but AsyncLocal correctly tracks current context and so it contains existing Activity instance (no new instance is created).

I suspect that other tests have the same problem and maybe the problem is not only in tests, but also in implementation.

**Comment by vladapokorny** [comment]
After more analysis, I found that the key part is ICreatedTransaction.GetForkScope. There is assumed that a new context is created (new Activity instance is created), which works with CallContext and ThreadLocal, but no with AsyncLocal. The solution for AsyncLocal is creating new Activity instance and storing it in AsyncLocal variable (value in original context remains unchanged). After this change, all unit tests are now passing.

**Comment by fir3pho3nixx** [comment]
> ICreatedTransaction.GetForkScope

Legendary catch! :)

So to release this:

 - We need a build system that runs tests against PR's.
 - We need a tagging system that deploys to NuGet via appveyor.
 - We need to harmonise the docs with castle in general(please follow README.md from Core/Windsor). 

I still thinking these things need to be looked at.

File: src/Castle.Transactions/Internal/TransactionLogic.cs
Comment (jonorossi) [inline src/Castle.Transactions/Internal/TransactionLogic.cs]: If `CreateNewActivity` returned the activity it created you wouldn't need to call `GetCurrentActivity` twice here.

**Comment by LightCZ** [comment]
Whats the current state of this PR? Is there any more work to do till it is considered to be merged?

**Comment by jonorossi** [comment]
What is the status of this pull request? It has been quiet nearly 8 weeks.

**Comment by LightCZ** [comment]
Would it be possible to merge this and not wait for AppVeyor build system. I dont see these two thing necessary related.

**Comment by jonorossi** [comment]
> Would it be possible to merge this and not wait for AppVeyor build system. I dont see these two thing necessary related.

No idea the state of this pull request. It appears to have an appveyor.yml file. We can't release packages without a successful automated build, that is a minimum quality standard.

**Comment by jonorossi** [comment]
Closing as stale for 18 months.

**Pull request closed without merging**
